
---
title: "IntLIM:  Integration through Linear Modeling"
author: "Jalal K. Siddiqui, Elizabeth Baskin, Mingrui Liu, Carmen Z. Cantemir-Stone, Bofei Zhang, Russell Bonneville, Joseph P. McElroy, Kevin R. Coombes, Ewy Mathé, Tara Eicher"
date: "3/29/2018"
output: rmarkdown::html_document
vignette: >
  %\VignetteIndexEntry{IntLIM:  Integration through Linear Modeling}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
result_dir <- "~\\IntLIM_vignette_results"
dir.create(result_dir)
```

##Introduction

Interpretation of metabolomics data is very challenging.  Yet it can be eased through integration of metabolomics with other ‘omics’ data. The IntLIM package, which includes a user-friendly RShiny web app, aims to integrate metabolomics data with transcriptomic data.  Unlike other approaches, IntLIM is focused on understanding how specific gene-metabolite associations are affected by phenotypic features.  To this end, we develop a linear modeling approach that describes how gene-metabolite associations are affected by phenotype.  More information can be found on <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-018-2085-6" target="_blank"> our publication "IntLIM: integration using linear models of metabolomics and gene expression data"</a>The workflow involves the following steps: 1) input gene expression/metabolomics data files, 2) filter data sets by gene and metabolite abundances and imputed values, 3) run the linear model to extract FDR-adjusted interaction p-values, 4) filter results by p-values and Spearman correlation differences, and 5) plot/visualize specific gene-metabolite associations. 

##Installation

IntLIM is an R package and can be run on version >= 3.2.0. 

Download (or upgrade) R here: https://cloud.r-project.org/

RStudio (an interface to R than can make R easier to use) can be download here (not required): https://www.rstudio.com/products/rstudio/download3/

Prior to installing IntLIM, it is necessary to have the the Bioconductor package *MultiDataSet* (Hernandez-Ferrer et al, 2017).  

The following command then installs *MultiDataSet*.  

```{r eval = FALSE}
## try http:// if https:// URLs are not supported
if (!requireNamespace("BiocManager", quietly = TRUE)){
  install.packages("BiocManager")
}
```

The function install_github() from the “devtools” package (Wickham and Chang, 2015) installs IntLIM directly from.  To install IntLIM, enter the following: 

```{r eval = FALSE}
if(!require("devtools")){
  install.packages("devtools")
}
library("devtools")
install_github("ncats/IntLIM@liz_dev")
```

IntLIM can be loaded using the library function

```{r}
library(IntLIM)
```

##Inputting Gene Expression and Metabolite Abundances

The first step is importing in the gene expression and metabolite data.  To this end we have provided a sample data file for users wishing to use IntLIM.  

IntLIM requires a specific format for gene expression and metabolite data.  For IntLIM, we require the metabolite data, gene expression data, metabolite meta data, gene meta data, and sample meta data.  We also need a CSV meta-file that lists the location of the other files.  These need to be in the same folder.  The formats are described below.  In addition, we provide a sample set of files.  

**Please be sure that all files noted in the CSV file, including the CSV file, are in the same folder. Do not include path names in the filenames.**

Users need to input a CSV file with two required columns: 'type' and 'filenames'.

The CSV file is expecected to have the following 2 columns and 6 rows:

1. type,filenames
2. metabData,myfilename
3. geneData,myfilename
4. metabMetaData,myfilename (optional)
5. geneMetaData,myfilename (optional)
6. sampleMetaData,myfilename"

The transcriptomic and metabolomic data and meta-data is stored in a series of comma-separated-values (.CSV) files.  The 5 files consist of normalized/log2-transformed metabolite abundance data, metabolite feature meta data, log2-transformed gene expression data, gene expression feature data, and sample meta data.  A meta-file lists the location of the other 5 fives.  This file is input into IntLIM.  These files should ideally be in the same folder.  Of note, input data files should be in a specific format: 

File type      | Description   
-------------- | -----------------------------------------
metabData      | rows are metabolites, columns are samples
geneData       | rows are genes, columns are samples 
metabMetaData  | rows are metabolites, features are columns 
geneMetaData   | rows are genes, features are columns
sampleMetaData | rows are samples, features are columns 

For the metabData and geneData files, the first row contains the feature ids and the first column contains the sample ids.  

For the sampleMetaData, the first column of the sampleMetaData file is assumed to be the sample id, and those sample ids should match the first row of metabData and geneData (e.g. it is required that all sample ids in the metabData and geneData are also in the sampleMetaDatafile).

Additionally, the metabMetaData, geneMetaData, and SampleMetaData need to contain an 'id' column that contains the name of the features (metabolite or gene) or sample (sample id, name, etc).

A small data set is embedded in a package.  This consists of the National Cancer Institute-60 (NCI-60) cell line data with a reduced number of genes for faster calculation.  To access it use the following commands.  csvfile describes the location of the meta-file describing the location of the other 5 input files.  

```{r}
dir <- system.file("extdata", package="IntLIM", mustWork=TRUE)
csvfile <- file.path(dir, "NCItestinput.csv")
csvfile
```

Through the __ReadData()__ function, users input the above .csv files containing normalized and log2-transformed gene expression data, normalized and log2-transformed metabolite abundance data, (optional) gene expression meta data, (optional) metabolite meta data, and sample meta data.  The gene expression data, meta data, and sample meta-data are input into an ExpressionSet object.  The metabolite abundance data, meta data, and sample data is input into a new MetaboliteSet object, a new eSet object designed to contain metabolomics data (Hernandez-Ferrer, et al., 2017).  Both objects are integrated into a MultiDataSet object- a multi-‘omics object allowing integration of eSet objects from different ‘omics data sets.  metabid and geneid describe the column name in the metabolite meta adata and the gene expression data to select as the identifer for that particular feature.  

```{r}
inputData <- IntLIM::ReadData(inputFile = csvfile,metabid='id',geneid='id')
```

The __ShowStats()__ function allows the user to summarize the metabolomic and gene expression data (how many total genes, metabolites, and samples in each data-set as well as samples in common).  We have a data-set that consists of 20 cell lines (samples), 1448 genes, and 257 metabolites. 

```{r}
IntLIM::ShowStats(IntLimObject = inputData)
```

##Filtering and Observing Data

Optionally, the __FilterData()__ function allows the user to filter out the features (genes or metabolites) based on their mean values. Users should input a percentile cutoff and any feature with a mean value below that cutoff will be removed.  Furthermore we can filter out metabolites by percentage of missing or imputed values.  For the analysis, we filter out the genes with the lowest 10% of gene expression and metabolites with more than 80% imputed values.  This is done as below. We henceforth have 1303 genes and 212 metabolites for 20 cell line samples.  

```{r}
inputDatafilt <- IntLIM::FilterData(inputData,geneperc = 0.10, metabmiss = 0.80)
IntLIM::ShowStats(inputDatafilt)
```

The __PlotDistribution()__ function allows users to produce a boxplot of the distribution of gene expression and metabolite abundance data.  This is done as below.   

```{r}
IntLIM::PlotDistributions(inputData = inputDatafilt)
```

Prior to running the model, the user can also perform a principal component analysis of the gene expression and metabolite data using the __PlotPCA()__ function.  The __stype__ command allows the user to select a column from the sample meta data that color-codes the samples into two categories (two cancer types, tumor vs. non-tumor, etc).  For the sample set, we select PBO (prostate/breast/ovarian cancers) vs. the leukemia group.  

```{r}
IntLIM::PlotPCA(inputData = inputDatafilt,stype = "PBO_vs_Leukemia")
```

##Run IntLIM
The linear models are run by the __RunIntLim()__ function.  The __stype__ command allows the user to select a column from the sample meta data for the two categories to be compared (two cancer types, tumor vs. non-tumor, etc).  Currently, IntLIM only supports comparison of two categories (__"PBO_vs_Leukemia"__ is selected). The resulting object from the analysis is an IntLimResults object containing slots for un-adjusted and False Discovery Rate (FDR)-adjusted p-values for the “(g:p)” interaction coefficient.  A significant FDR-adjusted p-value implies that the slope of gene-metabolite association in one phenotype is different from the other.  The __RunIntLim()__ function is based heavily on the MultiLinearModel functions developed for the ClassComparison package part of oompa (http://oompa.r-forge.r-project.org).

The following variations of this function are available:
- Setting `save.covar.pvals` to TRUE returns the p-values and coefficients of all terms in the model (not only the interaction term p-value and coefficient).
- Setting `remove.duplicates` to TRUE filters gene-gene and metabolite-metabolite association models to include only the model with the highest p-value. e.g. if m1 ~ m2 has a higher p-value than m2 ~ m1, then the m2 ~ m1 model will be returned. By default, only self-associations (i.e. m1 ~ m1) are removed.

```{r}
# Gene-metabolite
myres.gm <- IntLIM::RunIntLim(inputData = inputDatafilt,stype="PBO_vs_Leukemia", covar = c("cellnbr", "panelnbr"), class.covar = c("numeric", "numeric"), save.covar.pvals = TRUE, outcome = "metabolite", independent.var.type = "gene")

# Metabolite-gene
myres.mg <- IntLIM::RunIntLim(inputData = inputDatafilt,stype="PBO_vs_Leukemia", covar = c("cellnbr", "panelnbr"), class.covar = c("numeric", "numeric"), save.covar.pvals = TRUE, outcome = "gene", independent.var.type = "metabolite")

# Metabolite-metabolite
myres.mm <- IntLIM::RunIntLim(inputData = inputDatafilt,stype="PBO_vs_Leukemia", save.covar.pvals = TRUE, covar = c("cellnbr", "panelnbr"), class.covar = c("numeric", "numeric"), outcome = "metabolite", independent.var.type = "metabolite")

# Gene-gene
myres.gg <- IntLIM::RunIntLim(inputData = inputDatafilt,stype="PBO_vs_Leukemia", save.covar.pvals = TRUE, covar = c("cellnbr", "panelnbr"), class.covar = c("numeric", "numeric"), outcome = "gene", independent.var.type = "gene")
```

The __DistPValues()__ function allows the user to observe a histogram of the p-values prior to adjustment 

```{r}
IntLIM::DistPvalues(IntLimResults = myres.gm)
IntLIM::DistPvalues(IntLimResults = myres.mm)
IntLIM::DistPvalues(IntLimResults = myres.mg)
IntLIM::DistPvalues(IntLimResults = myres.gg)
```

The __DistRSquared()__ function allows the user to observe a histogram of the r-squared values.
```{r}
IntLIM::DistRSquared(IntLimResults = myres.gm)
IntLIM::DistRSquared(IntLimResults = myres.mm)
IntLIM::DistRSquared(IntLimResults = myres.mg)
IntLIM::DistRSquared(IntLimResults = myres.gg)
```

The __PValueBoxPlots()__ function allows the user to observe the p-value distributions of all covariates in the linear models prior to adjustment.
Note: To view these values, you must set __save.covar.pvals__ to __TRUE__ when running __RunIntLim__. 
```{r}
IntLIM::PValueBoxPlots(IntLimResults = myres.gm)
IntLIM::PValueBoxPlots(IntLimResults = myres.mm)
IntLIM::PValueBoxPlots(IntLimResults = myres.mg)
IntLIM::PValueBoxPlots(IntLimResults = myres.gg)
```

The __InteractionCoefficientGraph()__ function allows the user to observe the interaction coefficients of all models, sorted in ascending order.
```{r}
# Gene-metabolite
IntLIM::InteractionCoefficientGraph(inputResults = myres.gm, interactionCoeffPercentile = 0.9)

# Metabolite-metabolite
IntLIM::InteractionCoefficientGraph(inputResults = myres.mm, independent.var.type = "metabolite",
                                    outcome = "metabolite", interactionCoeffPercentile = 0.9)

# Metabolite-gene
IntLIM::InteractionCoefficientGraph(inputResults = myres.mg, independent.var.type = "metabolite",
                                    outcome = "gene", interactionCoeffPercentile = 0.9)

# Gene-gene
IntLIM::InteractionCoefficientGraph(inputResults = myres.gg, independent.var.type = "gene",
                                    outcome = "gene", interactionCoeffPercentile = 0.9)
```
The __MarginalEffectsGraph()__ function allows the user to observe the marginal effects of phenotype for a given gene and metabolite.
The __MarginalEffectsGraphDataframe()__ function creates the input for __MarginalEffectsGraph()__, given the IntLIM results, the input data, and the gene and metabolite of interest.
The __MarginalEffectsGraphDataframeMetabolitePairs()__ function is analogous to __MarginalEffectsGraphDataframe()__ for metabolite pairs.
```{r}
# Gene-metabolite
IntLIM::MarginalEffectsGraph(dataframe = IntLIM::MarginalEffectsGraphDataframe(inputResults = myres.gm, inputData = inputDatafilt, geneOfInterest = "DLG4", metaboliteOfInterest = "(p-Hydroxyphenyl)lactic acid"), title = "Marginal Effects - DLG4 and (p-Hydroxyphenyl)lactic acid")

# Metabolite-metabolite
IntLIM::MarginalEffectsGraph(dataframe = IntLIM::MarginalEffectsGraphDataframeMetabolitePairs(inputResults = myres.mm, inputData = inputDatafilt, metaboliteOfInterest1 = "alanine", metaboliteOfInterest2 = "(p-Hydroxyphenyl)lactic acid"), title = "Marginal Effects - alanine and (p-Hydroxyphenyl)lactic acid")

# Metabolite-gene
IntLIM::MarginalEffectsGraph(dataframe = IntLIM::MarginalEffectsGraphDataframe(inputResults = myres.mg, inputData = inputDatafilt, geneOfInterest = "DLG4", metaboliteOfInterest = "(p-Hydroxyphenyl)lactic acid"), title = "Marginal Effects - DLG4 and (p-Hydroxyphenyl)lactic acid")

# Gene-gene
IntLIM::MarginalEffectsGraph(dataframe = IntLIM::MarginalEffectsGraphDataframeGenePairs(inputResults = myres.gg, inputData = inputDatafilt, geneOfInterest1 = "DLG4", geneOfInterest2 = "ZRSR2"), title = "Marginal Effects - DLG4 and ZRSR2")
```

The __pvalCorrVolcano()__ function allows users to observe a volcano plot comparing the Spearman correlation difference between groups to the –log10(FDR-adjusted p-value).
The __pvalCorrVolcanoMetabolitePairs()__ function is analogous to the __pvalCorrVolcano__() function, but for metabolite pairs.
Note that these functions are not applicable to continuous outcome data.

```{r}
# Gene-metabolite
IntLIM::pvalCorrVolcano(inputResults = myres.gm, inputData = inputDatafilt, diffcorr = 0.5, pvalcutoff = 0.05)

# Metabolite-metabolite
IntLIM::pvalCorrVolcanoMetabolitePairs(inputResults = myres.mm, inputData = inputDatafilt, diffcorr = 0.5, pvalcutoff = 0.05)

# Metabolite-gene
IntLIM::pvalCorrVolcano(inputResults = myres.mg, inputData = inputDatafilt, diffcorr = 0.5, pvalcutoff = 0.05)

# Gene-gene
IntLIM::pvalCorrVolcanoGenePairs(inputResults = myres.gg, inputData = inputDatafilt, diffcorr = 0.5, pvalcutoff = 0.05)
```

Process the results and filter pairs of genes-metabolites based on adjusted p-values and differences in correlation coefficients between groups 1 and 2.  Then plot heatmap of significant gene-metabolite pairs

##Filter Results

The __ProcessResults()__ function allows the user to filter the results by FDR p-values (default set at 0.05) and by the absolute value difference of the gene-metabolite Spearman correlation (default set at 0.50) between the two groups.  The output is a list of gene-metabolite pairs and gene-metabolite Spearman correlations for each of the two groups
The __ProcessResultsMetabolitePairs()__ function is analogous to __ProcessResults()__ for metabolite pairs.
Note that the __diffcorr__ and __corrtype__ arguments must be set to __NULL__ for continuous phenotypes.

```{r}
# Gene-metabolite
myres.gm.sig <- IntLIM::ProcessResults(inputResults = myres.gm, inputData = inputDatafilt, pvalcutoff = 0.10, diffcorr = 0.5, rsquaredCutoff = 0.2)

# Metabolite-metabolite
myres.mm.sig <- IntLIM::ProcessResults(inputResults = myres.mm, inputData = inputDatafilt, pvalcutoff = 0.10, diffcorr = 0.5, rsquaredCutoff = 0.2)

# Metabolite-gene
myres.mg.sig <- IntLIM::ProcessResults(inputResults = myres.mg, inputData = inputDatafilt, pvalcutoff = 0.10, diffcorr = 0.5, rsquaredCutoff = 0.2)

# Gene-gene
myres.gg.sig <- IntLIM::ProcessResults(inputResults = myres.gg, inputData = inputDatafilt, pvalcutoff = 0.10, diffcorr = 0.5, rsquaredCutoff = 0.2)
```

The __CorrHeatmap()__ produces a clustered heatmap of the gene-metabolite Spearman correlations for each of these groups.  
Note: This function is only applicable to discrete phenotypes.

```{r}
IntLIM::CorrHeatmap(myres.gm.sig, inputDatafilt)
IntLIM::CorrHeatmap(myres.mm.sig, inputDatafilt)
IntLIM::CorrHeatmap(myres.mg.sig, inputDatafilt)
IntLIM::CorrHeatmap(myres.gg.sig, inputDatafilt)
```

The __HistogramGMPairs()__ function plots the count of gene-metabolite pairs by either
gene or metabolite.
The __HistogramMMPairs()__ function is the analogous function for metabolite pairs.
```{r}
IntLIM::HistogramGMPairs(myres.gm.sig, type = "gene")
IntLIM::HistogramGMPairs(myres.gm.sig, type = "metabolite")
IntLIM::HistogramMMPairs(myres.mm.sig, type = "outcome")
IntLIM::HistogramMMPairs(myres.mm.sig, type = "independent")
IntLIM::HistogramGMPairs(myres.mg.sig, type = "gene")
IntLIM::HistogramGMPairs(myres.mg.sig, type = "metabolite")
IntLIM::HistogramGGPairs(myres.gg.sig, type = "outcome")
IntLIM::HistogramGGPairs(myres.gg.sig, type = "independent")
```

##Visualize and Export Results

A __PlotGMPair()__ function  allows the user to plot a chosen gene-metabolite association for selected groups. An example is shown below for DLG4 vs. (p-Hydroxyphenyl)lactic acid.
The __PlotMMPair()__ function is analogous to the __PlotGMPair()__ function for metabolite pairs. An example is shown below for alanine vs. (p-Hydroxyphenyl)lactic acid.

Plot a pair of interest:

```{r}
IntLIM::PlotGMPair(inputDatafilt,stype="PBO_vs_Leukemia","DLG4","(p-Hydroxyphenyl)lactic acid")
IntLIM::PlotMMPair(inputDatafilt,stype="PBO_vs_Leukemia","alanine","(p-Hydroxyphenyl)lactic acid")
IntLIM::PlotMGPair(inputDatafilt,stype="PBO_vs_Leukemia","(p-Hydroxyphenyl)lactic acid","DLG4")
IntLIM::PlotGGPair(inputDatafilt,stype="PBO_vs_Leukemia","ZRSR2","DLG4")
```

Of note, plots are generated using Highcharter (http://jkunst.com/highcharter/) and Plotly (Sievert, et al.) (https://plot.ly), which enables interactive visualization, allowing users to promptly assess the effect of changing parameters on analysis results and accelerating discovery of phenotype-specific gene-metabolite pairs.  This will greatly allow the workflow to be accessible to non-computational biologists.   

The __OutputData()__ and __OutputResults()__ function allows the users to output the data and results of the analyses into zipped CSV files.  

Lastly, various writing functions are implemented.  The __OutputData()__ and __OutputResults()__ function allows the users to output the data and results of the analyses into zipped CSV files.

```{r eval = FALSE}
IntLIM::OutputData(inputData=inputDatafilt,filename=paste(result_dir, "FilteredData.zip", sep = "\\"))
IntLIM::OutputResults(inputResults=myres.gm.sig,filename=paste(result_dir, "MyResults.gm.csv", sep = "\\"))
IntLIM::OutputResults(inputResults=myres.mm.sig,filename=paste(result_dir, "MyResults.mm.csv", sep = "\\"))
IntLIM::OutputResults(inputResults=myres.mg.sig,filename=paste(result_dir, "MyResults.mg.csv", sep = "\\"))
IntLIM::OutputResults(inputResults=myres.gg.sig,filename=paste(result_dir, "MyResults.gg.csv", sep = "\\"))
```

# Create Co-Regulation Graph
In a co-regulation graph, nodes are analytes (genes and metabolites) and edges indicate
a significant phenotype-dependent association between the analytes.
```{r}
coreg_graph = BuildCoRegulationGraph(list(myres.gm.sig, myres.mm.sig, myres.mg.sig, myres.gg.sig))

# Betweenness
centrality_graph = IntLIM::ComputeCentrality(coreg_graph, c("betweenness", "kleinberg", "closeness",
                                                                      "eccentricity"))
write.csv(centrality_graph, file=paste(result_dir, "Graph.Centrality.csv", sep = "\\"))
```
# Run Pairwise Prediction.
```{r}
# Get predictions.
predictions_gm <- IntLIM::RunPairwisePrediction(myres.gm.sig, inputDatafilt, stype="PBO_vs_Leukemia", covar = c("cellnbr", "panelnbr"), class.covar = c("numeric", "numeric"))
predictions_mm <- IntLIM::RunPairwisePrediction(myres.mm.sig, inputDatafilt, stype="PBO_vs_Leukemia", covar = c("cellnbr", "panelnbr"), class.covar = c("numeric", "numeric"))
predictions_mg <- IntLIM::RunPairwisePrediction(myres.mg.sig, inputDatafilt, stype="PBO_vs_Leukemia", covar = c("cellnbr", "panelnbr"), class.covar = c("numeric", "numeric"))
predictions_gg <- IntLIM::RunPairwisePrediction(myres.gg.sig, inputDatafilt, stype="PBO_vs_Leukemia", covar = c("cellnbr", "panelnbr"), class.covar = c("numeric", "numeric"))

# Project predictions onto graph.
prediction_graphs <- IntLIM::ProjectPredictionsOntoGraph(list(predictions_gm, predictions_mm, predictions_mg, predictions_gg), coreg_graph)

# Plot predictions.
stype <- "PBO_vs_Leukemia"
lt_blue <- rgb(0,0,153,max = 255, alpha = 80, names = "lt.blue")
lt_pink <- rgb(153,0,0, max = 255, alpha = 80, names = "lt.pink")
lt_green <- rgb(0,153,0, max = 255, alpha = 80, names = "lt.green")
lt_gray <- rgb(64,64,64, max = 255, alpha = 80, names = "lt.gray")
for(samp in setdiff(colnames(predictions_gm), c("from", "to"))){
  gm_hist <- hist(as.numeric(predictions_gm[,samp]), breaks = 100, plot = FALSE)
  mm_hist <- hist(as.numeric(predictions_mm[,samp]), breaks = 100, plot = FALSE)
  mg_hist <- hist(as.numeric(predictions_mg[,samp]), breaks = 100, plot = FALSE)
  gg_hist <- hist(as.numeric(predictions_gg[,samp]), breaks = 100, plot = FALSE)
  min_x <- min(c(gm_hist$breaks, mm_hist$breaks, mg_hist$breaks, gg_hist$breaks))
  max_x <- max(c(gm_hist$breaks, mm_hist$breaks, mg_hist$breaks, gg_hist$breaks))
  max_y <- max(c(gm_hist$counts, mm_hist$counts, mg_hist$counts, gg_hist$counts))
  plot(gm_hist, xlab = paste("Predicted", stype), main = samp, xlim = c(min_x, max_x),
       ylim = c(0, max_y), ylab = "Number of Nodes with Prediction", col = lt_blue, 
       lty="blank")
  plot(mm_hist, xlab = "", main = "", ylab = "", col= lt_pink, lty="blank", 
       xlim = c(min_x, max_x), ylim = c(0, max_y), add = TRUE)
  plot(mg_hist, xlab = "", main = "", ylab = "", col= lt_green, lty="blank", 
       add = TRUE, xlim = c(min_x, max_x), ylim = c(0, max_y))
  plot(gg_hist, xlab = "", main = "", ylab = "", col = lt_gray, lty="blank", 
       add = TRUE, xlim = c(min_x, max_x), ylim = c(0, max_y))
  legend(x = min_x + (max_x - min_x) * 0.75, y = max_y * 0.75, legend = c("GM", "MM", "MG", "GG"),
         fill = c(lt_blue, lt_pink, lt_green, lt_gray))
}
```

```{r}
# Plot.
coreg_weightless <- igraph::as_adjacency_matrix(coreg_graph, sparse = FALSE)
coreg_weightless[which(coreg_weightless != 0)] <- 1
coreg_weightless_g <- igraph::graph_from_adjacency_matrix(coreg_weightless)
dist <- igraph::distances(coreg_weightless_g)
dist[which(!is.finite(dist))] <- NA
max_dist <- max(dist[which(!is.na(dist))])
dist[which(is.na(dist))] <- 2 * max_dist
hier <- stats::hclust(as.dist(dist), method = "average")
dend <- as.dendrogram(hier)
for(i in 1:length(prediction_graphs)){
  g <- prediction_graphs[[i]]
  adj <- igraph::as_adjacency_matrix(g, sparse = FALSE, attr = "weight")
  adj <- adj[rownames(coreg_weightless), colnames(coreg_weightless)]
  colors<- colorRampPalette(c("white", "red", "darkgreen"))(256)
  gplots::heatmap.2(adj, Colv=dend,Rowv=rev(dend),labRow=NA, density.info="none",
                    trace="none", col = colors)
}
```

# Train a graph learning based prediction model.
```{r}
# Format the input for training.
modelInputs <- IntLIM::formatTrainingInput(predictionGraphs=prediction_graphs, coregulationGraph=coreg_graph, inputData = inputDatafilt, stype = "PBO_vs_Leukemia", stype.class = "categorical")

# Compute distances between edges.
# g <- igraph::graph_from_adjacency_matrix(modelInputs@line.graph)
# dist <- igraph::distances(g)
# dist[which(!is.finite(dist))] <- dim(dist)[1] * 2
# diffs <- lapply(1:ncol(modelInputs@node.wise.prediction), function(i){
#   predictions <- modelInputs@node.wise.prediction[,i]
#   differences <- matrix(0, nrow(dist), ncol(dist))
#   rownames(differences) <- rownames(dist)
#   colnames(differences) <- colnames(dist)
#   for(j in 1:length(rownames(dist))){
#     r <- rownames(dist)[j]
#     pred_r <- predictions[r]
#     for(c in colnames(dist)){
#       pred_c <- predictions[c]
#       differences[r,c] <- abs(pred_r - pred_c)
#     }
#   }
#   print(colnames(modelInputs@node.wise.prediction)[i])
#   return(differences)
# })
```

```{r}
# Make plots.
# for(i in 1:ncol(modelInputs@node.wise.prediction)){
#   diff <- diffs[[i]]
#   samp <- colnames(modelInputs@node.wise.prediction)[i]
#   diff_which_adjacent <- as.vector(diff[which(dist == 1)])
#   diff_which_not_adjacent <- as.vector(diff[which(dist > 1)])
#   h_adj <- hist(diff_which_adjacent, breaks = 1000, plot = FALSE)
#   h_nadj <- hist(diff_which_not_adjacent, breaks = 1000, plot = FALSE)
#   plot(h_adj, col = rgb(173,216,230,max = 255, alpha = 80, names = "lt.blue"), lty="blank", main = samp, ylab = "Pair Count", xlab = "Prediction Difference Between Linear Model Pairs")
#   plot(h_nadj, col = rgb(255,192,203, max = 255, alpha = 80, names = "lt.pink"), lty="blank", main = "", ylab = "", xlab = "", add = TRUE)
#   text(x = max(diff_which_adjacent)*0.75, y = max(h_adj$counts)*0.75, paste("Wilcoxon Test p-Value", formatC(wilcox.test(diff_which_adjacent, diff_which_not_adjacent, alternative = "less")$p.value, format = "e", digits = 2)))
# }
```

```{r}
# Make plots.
# dist <- igraph::distances(g)
# dist[which(!is.finite(dist))] <- max(dist) * 2
# for(i in 1:ncol(modelInputs@node.wise.prediction)){
#   diff <- diffs[[i]]
#   samp <- colnames(modelInputs@node.wise.prediction)[i]
#   dist_vec <- as.vector(dist)
#   diff_vec <- as.vector(diff)
#   plot(diff_vec, dist_vec, col = rgb(173,216,230,max = 255, alpha = 80, names = "lt.blue"), lty="blank", main = samp, ylab = "Graph Distance", xlab = "Prediction Difference")
# }
```

```{r}
# Create pooling filter.
hier <- IntLIM::doHierarchicalClustering(modelInputs = modelInputs)

# Create the dendrogram.
dend <- as.dendrogram(hier)
dendrogram_data <- ggdendro::dendro_data(dend)
dendrogram_segments <- dendrogram_data$segments
dendrogram_ends <- dplyr::filter(dendrogram_segments, yend == 0)
dendrogram_ends <- dplyr::left_join(filtered_yend, dendrogram_data$labels, by = "x")
dendrogram_ends <- dplyr::rename(joined_by_x, node_name = label)

# Plot each dendrogram with predictions.
plot_dendrogram = function(i){
  # Add predictions to dendrogram.
  all_pred <- dendrogram_ends
  all_pred$pred <- modelInputs@node.wise.prediction[dend_info$node_name, i]
  
  # Make the plot.
  ggplot2::ggplot() + 
    geom_segment(data = dendrogram_segments, aes(x=x, y=y, xend=xend, yend=yend)) +
    geom_segment(data = all_pred, aes(x=x, y=y.x, xend=xend, yend=yend, color=pred)) +
    scale_color_gradient(low = "red", high = "blue")+
    scale_y_reverse() +
    coord_flip() +
    theme_bw() +
    theme(panel.border = element_blank(), panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(), axis.title.x=element_blank(),
          axis.title.y=element_blank(), axis.text.x=element_blank(),
          axis.text.y=element_blank(), axis.ticks.x=element_blank(),
          axis.ticks.y=element_blank()) +
    labs(color='Prediction') +
    ggtitle(colnames(modelInputs@node.wise.prediction)[i])
}
for(i in 1:length(prediction_graphs)){
  plot_dendrogram(i)
}
```

```{r}
var <- IntLIM::findHierarchicalClusteringVariance(hier)
poolingFilter <- IntLIM::CreatePoolingFilterKMeans(modelInputs = modelInputs, k = length(prediction_graphs), poolType = "median")

# Plot.
# for(i in 1:length(prediction_graphs)){
#   line_graph <- modelInputs@line.graph
#   line_graph[which(line_graph != 0)] <- 0
#   for(clust in poolingFilter){
#     line_graph[clust[[1]], clust[[1]]] <- mean(modelInputs@node.wise.prediction[clust[[1]]
#                                                                                 ,i])
#   }
#   gplots::heatmap.2(line_graph, density.info="none", trace="none", col = col)
# }

# Initialize model.
modelResultsInit <- IntLIM::InitializeGraphLearningModel(modelInputs, poolingFilter,
                                                     iterations = 10000,
                                                     convergenceCutoff = 0.0001,
                                                     learningRate = 0.9,
                                                     activationType = "sigmoid",
                                                     optimizationType = "BGD",
                                                     weightsAfterPooling = FALSE)

# Train model.
modelResultsFinal <- IntLIM::TrainGraphLearningModel(modelResults = modelResultsInit)
#PlotGraphWeightsHeatmap(coreg_graph, modelResultsFinal)
#write.csv(modelResultsFinal@iteration.tracking,paste(result_dir, "model.results.csv", sep = "\\"))
```

##References

Siddiqui JK, Baskin E, Liu M, Cantemir-Stone CZ, Zhang B, Bonneville R, McElroy JP, Coombes KR, Mathé EA. IntLIM: integration using linear models of metabolomics and gene expression data. BMC bioinformatics. 2018 Dec;19(1):81.

Hernandez-Ferrer, C., et al. MultiDataSet: an R package for encapsulating multiple data sets with application to omic data integration. BMC bioinformatics 2017;18(1):36.

Sievert, C., et al. plotly: Create Interactive Web Graphics via’plotly. js’. 2016. R package version 3.6. 0. In.

Wickham, H. and Chang, W. devtools: Tools to make developing R code easier. R package version 2015;1(0).
